### ISSUE TITLE
KubernetesClient sends volumeName as empty string for PersistentVolumeClaim causing Forbidden: spec is immutable after creation

### ISSUE THREAD

**Issue by uklance** [fabric8io/kubernetes-client#4332]
I want to apply the following yaml multiple times with the [fabric8 kubernetes-client](https://github.com/fabric8io/kubernetes-client)

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-pvc
spec:
  storageClassName: my-storage-class
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
```

I apply the yaml using `createOrReplace()`
```java
Config config = new ConfigBuilder()
		.withMasterUrl("https://my-kubernetes-root:6443")
		.withNamespace("my-namespace")
		.withOauthToken(token)
		.withTrustCerts(true)
		.build();

KubernetesClient client = new DefaultKubernetesClient(config);
ClasspathResource resource = new ClasspathResource("my-pvc.yaml");

client.load(resource.getInputStream()).createOrReplace(); // this works
TimeUnit.MINUTES.sleep(1); // volumeName is dynamically assigned during this period
client.load(resource.getInputStream()).createOrReplace(); // this fails
```

**This works the first time** (when the PVC does not exist in the namespace) **but fails the second time** that `createOrReplace()` is called for the same yaml with the following error
```
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: https://my-kubernetes-root:6443/api/v1/namespaces/my-namespace/persistentvolumeclaims/my-pvc. Message: PersistentVolumeClaim "my-pvc" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims
  core.PersistentVolumeClaimSpec{
  	AccessModes:      []core.PersistentVolumeAccessMode{"ReadWriteMany"},
  	Selector:         nil,
  	Resources:        core.ResourceRequirements{Requests: core.ResourceList{s"storage": {i: resource.int64Amount{value: 1073741824}, s: "1Gi", Format: "BinarySI"}}},
- 	VolumeName:       "",
+ 	VolumeName:       "pvc-b79ebfcb-d5cb-4450-9f17-d69ec10b8712",
  	StorageClassName: &"my-storage-class",
  	VolumeMode:       &"Filesystem",
  	DataSource:       nil,
  }
```

Notice how **"volumeName" is not present in the yaml** (nil) but in the error message **"volumeName" is changing from empty string** to the dynamically assigned volumeName.

**I can reproduce this exact same behavior using kubectl and empty string for volumeName**

I can `kubectl apply` the following yaml as many times as I like
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-pvc
spec:
  storageClassName: my-storage-class
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
```

But if I `kubectl apply` a yaml with volumeName of empty string **it works the first time and fails the second time** (The error message is the same as above)
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-pvc
spec:
  storageClassName: my-storage-class
  volumeName: ""
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
```

How can I stop KubernetesClient from sending the volumeName as empty string? Or is there any other way to fix this so I can apply the same PersistentVolumeClaim yaml multiple times?

I've asked this same issue on stack overflow [KubernetesClient sends volumeName as empty string for PersistentVolumeClaim causing Forbidden: spec is immutable after creation](https://stackoverflow.com/questions/73330768/fabric8-kubernetesclient-sends-volumename-as-empty-string-for-persistentvolumecl)

**Comment by shawkins** [comment]
The problem is the replace operation.  It is not a substitute for apply / patch.  Instead of createOrReplace you can use the new support for server side apply - https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#server-side-apply

Or you can use separate create and patch calls.

**Comment by uklance** [comment]
Currently I have multiple entities in a single yaml file (seperated by "---") and I can `createOrReplace(...)` them all in a single operation.

```
KubernetesClient client = new DefaultKubernetesClient(config);
ClasspathResource resource = new ClasspathResource("multiple-entities.yaml");
client.load(resource.getInputStream()).createOrReplace();
```

I don't seem to have this option with serverside apply. I was hoping I could do
```
KubernetesClient client = new DefaultKubernetesClient(config);
ClasspathResource resource = new ClasspathResource("multiple-entities.yaml");
client.load(resource.getInputStream()).patch(PatchContext.of(PatchType.SERVER_SIDE_APPLY);
```

But I can't . What are my options for doing a server side apply for multiple entities (deployment, service, PVC etc) in a single yaml?

**Comment by shawkins** [comment]
There isn't a top level operation on a resource list for patching, but 6.0 has added the resources method to help process composite operations:

```
client.load(resource.getInputStream()).resources().forEach(r -> r.patch(PatchContext.of(PatchType.SERVER_SIDE_APPLY)));
```

**Comment by uklance** [comment]
That sounded like it would work but unfortunately `resources()` method is not available. 

[KubernetesClient.load(InputStream)](https://javadoc.io/doc/io.fabric8/kubernetes-client/5.12.3/io/fabric8/kubernetes/client/KubernetesClient.html#load-java.io.InputStream-) returns [ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable](https://javadoc.io/doc/io.fabric8/kubernetes-client/5.12.3/io/fabric8/kubernetes/client/dsl/ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable.html) which does not have a `resources()` method.

**Comment by uklance** [comment]
Ah, I hadn't realized that 6.0 was released (last time I checked it was RC). Thanks for the info, there seem to be a few changes in 6.0 so I will need to fix my compile errors with 6.0 before verifying this works
